### DISCUSSION TITLE
Comparison with Typescript (tsc)

### DISCUSSION THREAD

**Discussion by xunilrj** [rome/tools#1888] [General]
These are the differences found so far between tsc and our parser.
I will be updating the section below as I find new differences.

# Differences

## JS_ARRAY_HOLE

tsc creates an empty node where it started to search for an item, but found the "hole". We put it after the trivia of last token ("[" or ",").

.\test262\test\built-ins\Array\from\from-array.js

```javascript
...
var array = [0, 'foo', , Infinity];
            ^         ^
           211       221
...
```

tsc

```
...
4....ArrayLiteralExpression@211..233
5.....NumericLiteral@212..213
5.....StringLiteral@215..220
5.....OmittedExpression@221..221
5.....Identifier@224..232
...
```

our parser

```
...
7.......JS_ARRAY_EXPRESSION@211..233
8........L_BRACK@211..212
8........LIST@212..232
9.........JS_NUMBER_LITERAL_EXPRESSION@212..213
10..........JS_NUMBER_LITERAL@212..213
9.........COMMA@213..215
9.........JS_STRING_LITERAL_EXPRESSION@215..220
10..........JS_STRING_LITERAL@215..220
9.........COMMA@220..222
9.........JS_ARRAY_HOLE@222..222
9.........COMMA@222..224
9.........JS_IDENTIFIER_EXPRESSION@224..232
10..........JS_REFERENCE_IDENTIFIER@224..232
8........R_BRACK@232..233
...
```

## JS_SEQUENCE_EXPRESSION

```javascript
...
(0,eval)(
^      ^
1092 1100
  '{ function f() { initialBV = f; f = 123; currentBV = f; return "decl"; } }'
);
...
```

tsc

```
...
4....BinaryExpression@1093..1099
5.....NumericLiteral@1093..1094
5.....CommaToken@1094..1095
5.....Identifier@1095..1099
...
```

our

```
...
5.....JS_SEQUENCE_EXPRESSION@1093..1099
6......JS_NUMBER_LITERAL_EXPRESSION@1093..1094
7.......JS_NUMBER_LITERAL@1093..1094
6......COMMA@1094..1095
6......JS_IDENTIFIER_EXPRESSION@1095..1099
7.......JS_REFERENCE_IDENTIFIER@1095..1099
8........IDENT@1095..1099
...
```

**Comment by ematipico** [comment]
The more I see this comparison, the more I am unsure about its usefulness. 

I may understand that we want to make sure that our AST is valid because we are building from scratch, although I also think that every parser would follow undert this situation. We will create the AST based on our needs and those needs might be different from TS (or any other parser). Is it worth keep exploring it?

For example, if we decided to create a CSS parser, we wouldn't have a decent comparison because the current state of CSS parsers is not that great (what we were doing was good quality). 

So my question stays: is it worth it?

**Reply by xunilrj** [reply]
It was more work than I initially thought; and we did find some problems.
So, I would not put this as super high priority, but I vote to continue this later.

**Comment by MichaReiser** [comment]
#1977 should fix the issue with the sequence expression parsing
